Build-process.md

```diff
@@ -0,0 +1,114 @@
+4. `app.yaml` is generated from `app_dev.yaml`. The two files are identical in plain prod mode, except for a comment in `app.yaml` noting that it is auto-generated.
+5. The third-party libraries, assets, files compiled by webpack, and some other files are all copied to a build folder `build/`. This mimics the build folder we generate and upload to production servers when deploying the app for real. During this process, asset files are renamed to include the hashes we generated earlier. This ensures that if the file content changes, the name changes, so any caches of the old file are invalidated.
+
+This is the mode our tests run in (except the lighthouse accessibility tests), and it's what you'll want to use if you want to locally run a version of the app that's as close to the production version as possible. You can use this mode like this:
+
+```console
+python -m scripts.start --prod_env
+```
+
+#### Minify only third-party libraries
+
+In this mode, the third-party libraries are installed and minified, but no other steps from [plain prod mode](#plain-prod-mode) happen. This is only used by the frontend tests.
+
+#### Maintenance mode
+
+As far as the build process is concerned, maintenance mode works just like [plain prod mode](#plain-prod-mode) except that the `ENABLE_MAINTENANCE_MODE` constant gets set to `true`. Then when the app runs, only admins can log in. We use this when we are upgrading the production server and need to let jobs run while ensuring users don't change any data.
+
+This mode is used by our deployment scripts. You can also enable it locally like this:
+
+```console
+python -m scripts.start --prod_env --maintenance_mode
+```
+
+#### Deploy mode
+
+Deploy mode is also very similar to [plain prod mode](#plain-prod-mode), except that the `EMULATOR_MODE` constant is set to `false`. Further, we make some changes to the `app.yaml` file:
+
+* Remove the `version: default` line
+* Remove the environment variables specified in the `ENV_VARS_TO_REMOVE_FROM_DEPLOYED_APP_YAML` constant in `build.py`. We remove these since they are already present in the production environment.
+
+This mode is used by our deployment scripts.
+
+## Source maps
+
+Even when running in dev mode, the code your browser runs does not look much like the code on your file system. Files have been combined, and in prod mode, they've been minified. To make debugging easier, you can generate source maps that map from the code your browser is running back to the code on your file system.
+
+Generating source maps is expensive, so by default we don't make them. However, you can enable source map generation in both dev and prod modes, usually by passing a `--source_maps`, for example:
+
+```console
+python -m scripts.start --source_maps
+python -m scripts.run_e2e_tests --prod_env --source_maps
+```
+
+With these flags, the `webpack.*.sourcemap.config.ts` configuration files are used to enable source mapping.
```

Comment:
Maybe include some explanation on how to use the source maps once they're generated (or link to a page explaining that)? (This also seems like something that should be covered or linked to on a "debugging tips/guide" page.)

Reply:
done

Reply:
> (This also seems like something that should be covered or linked to on a "debugging tips/guide" page.)

Any thoughts on / plans for this bit?

Reply:
This made me realize that while we have guides for debugging tests, we don't have any guides for debugging the Oppia code directly. This would probably be most useful for debugging the frontend code with the Chrome dev tools. I added #39 to track adding a page for that